src/services/document_service.py

```diff
@@ -454,78 +362,91 @@ def _merge_entity_and_files(node: Node, files: Dict[str, BinaryIO]):
+        if document.get("type") != SIMOS.PACKAGE.value and not document.get("isRoot"):
+            raise BadRequestException("Only root packages may be added without a parent.")
+
+        new_node = tree_node_from_dict(
+            document, blueprint_provider=self.get_blueprint, recipe_provider=self.get_storage_recipes
+        )
+
+        try:
+            if self.get_document(f"/{data_source_id}/{new_node.name}", resolve_links=True, depth=99):
+                raise ValidationException(
+                    message=f"A root package named '{new_node.name}' already exists",
+                    data={"dataSource": data_source_id, "document": document},
+                )
+        except NotFoundException:
+            pass
+
+        new_node.set_uid()
+
+        self.save(new_node, data_source_id, update_uncontained=update_uncontained)
+
+        return {"uid": new_node.node_id}
+
     def add(
         self,
-        data_source_id: str,
-        path: str | None,
-        document: Entity,
+        reference: str,
+        document: dict,
         files: dict[str, BinaryIO],
         update_uncontained=False,
     ):
         """Add en entity to path
-        path: dotted path on format 'RootPackage/folder/entity.attribute.attribute.
-            If none, we're adding to the data source itself.
+        reference: reference to a package or a data source.
         document: The entity to be added
         files: Dict with names and files of the files contained in the document
         update_uncontained: Whether to update uncontained children
         """
-        document_dict = document.dict()
-        validate_entity_against_self(document_dict, self.get_blueprint)
-
-        if not path:  # We're adding something to the dataSource itself
-            if not document.type == SIMOS.PACKAGE.value or not document_dict.get("isRoot", False):
-                raise BadRequestException("Only root packages may be added to the root of a data source")
-            try:
-                if self.get_document(f"/{data_source_id}/{document_dict['name']}", resolve_links=True, depth=99):
-                    raise ValidationException(
-                        message=f"A root package named '{document_dict['name']}' already exists",
-                        data={"dataSource": data_source_id, "document": document_dict},
-                    )
-            except NotFoundException:
-                pass
-            document_repository = self.repository_provider(data_source_id, self.user)
-            document_repository.update(document_dict)
-            return {"uid": document_dict["_id"]}
+        validate_entity_against_self(document, self.get_blueprint)
+        entity: Entity = Entity(**document)
 
-        target: Node = self.get_document(f"/{data_source_id}/{path}", resolve_links=True, depth=99)
+        if "/" not in reference:  # We're adding something to the dataSource itself
```

Comment:
The split_data_source_and_reference crashes if the references only contain a reference to the data source.

But this can be a good idea, I will add it to the to-do list for later, but not include it in this PR to make it bigger.